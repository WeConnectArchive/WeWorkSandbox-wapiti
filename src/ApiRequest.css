.title{font-family:var(--font-regular);font-size:var(--title-font-size);font-weight:bold;margin-bottom:16px}.param-name,.param-type{margin:1px 0;text-align:right;line-height:12px}.param-name{color:var(--fg);font-family:var(--font-mono)}.param-type{color:var(--light-fg);font-family:var(--font-regular)}.param-constraint{min-width:100px}.param-constraint:empty{display:none}.top-gap{margin-top:24px}.tab-buttons{height:30px;border-bottom:1px solid var(--light-border-color);align-items:stretch}.tab-btn{border:none;background-color:transparent;cursor:pointer;outline:none;font-size:var(--small-font-size);margin-right:16px;padding:1px}.tab-btn.active{border-bottom:3px solid var(--primary-color);font-weight:bold;color:var(--primary-color)}.tab-btn:hover{color:var(--primary-color)}.tab-content{margin:-1px 0 0 0}.link{font-size:var(--small-font-size);text-decoration:underline;color:var(--link-color);font-family:var(--font-mono);margin-bottom:2px}.textarea{min-height:180px;padding:5px;resize:vertical;font-size:var(--font-mono-size)}.response-message.error{color:var(--error-color);font-weight:bold;text-overflow:ellipsis}.response-message.success{color:var(--success-color);font-weight:bold;text-overflow:ellipsis}@media only screen and (min-width: 768px){.textarea{padding:16px}}
