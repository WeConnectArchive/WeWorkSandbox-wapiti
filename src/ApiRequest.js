import React, {Component} from 'react';
import cx from 'classnames';
import marked from 'marked';
import DangerousHTML from 'react-dangerous-html';
import {schemaToModel, getTypeInfo,  generateExample, removeCircularReferences} from './utils/common-utils.js';
import {TagInput} from './TagInput';
import {SchemaTree} from './SchemaTree';

import style from './ApiRequest.scss';
import stylee from './EndPoint.scss';
import stylefo from './FontStyles.scss';
import stylet from './TableStyles.scss';
import stylett from './TabStyles.scss';
import stylef from './FlexStyles.scss';
import stylei from './InputStyles.scss';

export class ApiRequest extends Component {

  constructor(props) {
    super(props);
    this.state = { exampleTabActive: true, mimeTypeSelected: 'none' };
  }

  render() {
    return (
    <div className={cx(stylee["request"],stylef["col"], stylefo["regular-font"], style["request-panel"])}>
      <div className={cx(style["title"])}>REQUEST</div>
      {this.inputParametersTemplate('path')}
      {this.inputParametersTemplate('query')}
      {this.requestBodyTemplate()}
      {this.inputParametersTemplate('header')}
      {this.inputParametersTemplate('cookie')}
    </div>
   );
  }

  inputParametersTemplate(paramType){
    let title ="";
    let filteredParams= this.props.parameters? this.props.parameters.filter( param => param.in === paramType ):[];
    if (filteredParams.length === 0 ){
      return "";
    }
    if (paramType==='path'){ title = "PATH PARAMETERS"}
    else if (paramType==='query'){ title = "QUERY-STRING PARAMETERS"}
    else if (paramType==='header'){ title = "REQUEST HEADERS"}
    else if (paramType==='cookie'){ title = "COOKIES"}


    const tableRows = [];
    for (const param of filteredParams) {
      if (!param.schema){
        continue;
      }
      let paramSchema = getTypeInfo(param.schema);
      if (paramSchema === undefined) return;
      let inputVal='';
      if (param.example){
        if (param.example==='0' || param.example === 0){
          inputVal='0'
        }
        else{
          inputVal = param.example;
        }
      }
      else{
         inputVal = paramSchema.default;
      }

      tableRows.push((<React.Fragment>
      <tr>
        <td style={{ minWidth: '100px'}}>
          <div className={cx(style["param-name"])}>
            {param.required && <span style={{ color: 'orangered' }}>*</span>}{param.name}
          </div>
          <div className={cx(style["param-type"])}>
          {paramSchema.type==='array' ? paramSchema.array
          : paramSchema.type + (paramSchema.format ? '\u00a0' + paramSchema.format : '')}
          </div>
        </td>
        <td style={{ minWidth: '100px'}}>
          {paramSchema.type === 'array' ? (
          <TagInput
            theClassName="request-param"
            theStyle={{ width: '100%', fontSize: 'calc(var(--small-) + 1px)', background: 'var(--input-bg)', lineHeight: '13px' }}
            dataPtype={paramType}
            dataPname={param.name}
            dataArray="true"
            placeholder="add-multiple\u23ce"
          ></TagInput>)
          :(<input type="text" spellCheck={false} style={{ width: '100%' }} className={cx(style["request-param"])}
            value={inputVal}/>)}
        </td>
        <td>
          <div className={cx(style["param-constraint"])}><div>
            {paramSchema.constrain && (<div>{paramSchema.constrain}<br/></div>)}
            {paramSchema.allowedValues && (<div>{paramSchema.allowedValues}</div>)}
          </div></div>
        </td>
      </tr>
      {param.description && (
        <tr>
          <td style={{ border: 'none' }}>

          </td>
          <td colSpan={2} style={{ border: 'none', marginTop: '0', padding: '0 5px' }}>
            <span className={cx(stylefo["m-markdown-small"])}><DangerousHTML html={marked(param.description)} /></span>
          </td>
        </tr>
        )}
       </React.Fragment>
      ))
    }

    return (<React.Fragment>
    <div className={cx(stylet["table-title"], style["top-gap"])}>{title}</div>
    <div style={{ display: 'block', overflowX: 'auto',  maxWidth: '100%' }}>
      <table className={cx(stylet["m-table"])} style={{ width: '100%', wordBreak: 'break-word' }}>
        {tableRows}
      </table>
    </div>
    </React.Fragment>
    )
  }

  activateTab = (e) => {
    this.setState({ exampleTabActive: e.target.id === 'tab_example', mimeTypeSelected: this.state.mimeTypeSelected });
  }

  onMimeTypeChange = (e) => {
    this.setState({ exampleTabActive: this.state.exampleTabActive, mimeTypeSelected: e.target.value });
  }

  requestBodyTemplate(){

    if(!this.props.requestBody){
      return '';
    }
    if (Object.keys(this.props.requestBody).length === 0){
      return '';
    }

    let mimeReqCount = 0;
    let shortMimeTypes = {};
    let bodyDescrHtml = this.props.requestBody.description? (<div className={stylefo["m-markdown"]}> <DangerousHTML html={marked(this.props.requestBody.description)} /></div>):'';
    let textareaExampleHtml = [];
    var formDataHtml;
    const formDataTableRows = [];
    let isFormDataPresent = false;
    let reqSchemaTree = "";

    let content = this.props.requestBody.content;

    for(let mimeReq in content ) {
      if (this.state.mimeTypeSelected === 'none') {
        this.setState({mimeTypeSelected: mimeReq, exampleTabActive: this.state.exampleTabActive })
      }
      // do not change shortMimeTypes values, they are referenced in other places
      if (mimeReq.includes('json')){shortMimeTypes[mimeReq]='json';}
      else if (mimeReq.includes('xml')){shortMimeTypes[mimeReq]='xml';}
      else if (mimeReq.includes('text/plain')){shortMimeTypes[mimeReq]='text';}
      else if (mimeReq.includes('form-urlencoded')){shortMimeTypes[mimeReq]='form-urlencoded';}
      else if (mimeReq.includes('multipart/form-data')){shortMimeTypes[mimeReq]='multipart-form-data';}

      let mimeReqObj = content[mimeReq];

      if (mimeReq.includes('json') || mimeReq.includes('xml') || mimeReq.includes('text/plain')){
        try {
          //Remove Circular references from RequestBody json-schema
          mimeReqObj.schema = JSON.parse(JSON.stringify(mimeReqObj.schema, removeCircularReferences()));
        }
        catch (err){
          console.error("Unable to resolve circular refs in schema", mimeReqObj.schema);
          return;
        }
        reqSchemaTree = schemaToModel(mimeReqObj.schema,{});
        let reqExample    = generateExample(
          mimeReqObj.schema? mimeReqObj.schema.examples:'',
          mimeReqObj.schema? mimeReqObj.schema.example:'',
          mimeReqObj.schema,
          mimeReq, "text"
        );
        textareaExampleHtml[mimeReq] = (<textarea
            className={cx(style['textarea'], stylei['textarea'], stylei['mono'], stylei['request-body-param'], shortMimeTypes[mimeReq])}
            spellCheck={false}
            style={{ resize: 'vertical', display: shortMimeTypes[mimeReq]==='json'?'block':'none' }}
            value={reqExample[0].exampleValue} />)
      }
      else if (mimeReq.includes('form') || mimeReq.includes('multipart-form')){
        isFormDataPresent = true;
        for (const fieldName in mimeReqObj.schema.properties)  {
          const fieldSchema = mimeReqObj.schema.properties[fieldName];
          const fieldType = fieldSchema.type;
          const arrayType = fieldSchema.type==='array'?fieldSchema.items.type:'';
          formDataTableRows.push((<div>
          <tr>
            <td style={{minWidth: '100px' }}>
              <div className={cx(style["param-name"])}>{fieldName}</div>
              <div className={cx(style["param-type"])}>
              {fieldType==='array' ? '{fieldType} of {arrayType}'
              : fieldType + (fieldSchema.format ? '\u00a0' + fieldSchema.format : '')}
              </div>
            </td>
            <td style={{ minWidth: '100px' }}>
              {fieldType === 'array'? (
              <TagInput 
                theClassName="request-form-param"
                theStyle={{ width: '100%', fontSize: 'calc(var(--title-font-size) + 1px)', background: 'var(--input-bg)', lineHeight: '13px' }}
                dataPtype={fieldType}
                dataPname={fieldName}
                dataArray="true"
                placeholder="add-multiple\u23ce"
              ></TagInput>)
              :(<input
                spellCheck={false}
                type="{fieldSchema.format==='binary'?'file':'text'}"
                style={{ width: '100%' }} className={cx(style["request-form-param"])}
                />
              )}

            </td>
            <td>
              <div className={cx(style["param-constraint"])}></div>
            </td>
          </tr>
          {fieldSchema.description?(
            <tr>
              <td style={{ border: 'none' }}></td>
              <td colSpan={2} style={{ border: 'none', marginTop:0, padding: '0 5px' }}>
                <span className={cx(stylec["m-markdown-small"])}><DangerousHTML html={marked(fieldSchema.description)} /></span>
              </td>
            </tr>):''}
           </div>
        ))
        }


        formDataHtml = (
        <form className={cx(style["{shortMimeTypes[mimeReq]}"])}> {/* onSubmit={{ event.preventDefault() }}> */}
          <table style={{ width: '100%' }} className={cx(stylet["m-table"])}>
            {formDataTableRows}
          </table>
        </form>);
      }
      mimeReqCount++;
    }

    let bodyTitle = "";
    if (this.props.specType === 'grpc') {
      bodyTitle = this.props.requestBody.objType
      textareaExampleHtml['protobuf'] = (<textarea
            className={cx(style['textarea'],stylei['textarea'],stylei["mono"], stylei["request-body-param"], "json")}
            spellCheck={false}
            style={{ resize: 'vertical', display: 'block' }}
            value={JSON.stringify(this.props.requestBody.mock, undefined, 2)} />)
      reqSchemaTree = this.props.requestBody.schema;
      shortMimeTypes['protobuf']={};
      if (this.state.mimeTypeSelected === 'none') {
        this.setState({mimeTypeSelected: 'protobuf', exampleTabActive: this.state.exampleTabActive })
      }
      mimeReqCount = 1;
    } else {
      bodyTitle = (isFormDataPresent?'FORM':'BODY') + ' DATA '+(this.props.requestBody.required?'(required)':'')
    }

    let styleExample=this.state.exampleTabActive?{ flex: 1 }:{ flex: 1, display: 'none' }
    let styleModel=this.state.exampleTabActive?{ flex: 1, display: 'none' }:{ flex: 1 }

    return (<React.Fragment>
      <div className={cx(stylet['table-title'], style['top-gap'], style[(isFormDataPresent?'form_data':'body_data')])}> {bodyTitle} </div>
      {bodyDescrHtml}
      {isFormDataPresent?({formDataHtml})
        :(
        <div className={cx(stylett["tab-panel"], stylef["col"])} style={{borderWidth: '0',  minHeight: '200px' }}>
          <div id="tab_buttons" className={cx(stylett["tab-buttons"], stylef["row"])} onClick={this.activateTab}>
            <button className={cx(style['tab-btn'], style[(this.state.exampleTabActive?'active':'')])} id="tab_example">EXAMPLE </button>
            <button className={cx(style['tab-btn'], style[(this.state.exampleTabActive?'':'active')])} id="tab_model">MODEL</button>
            <div style={{ flex:1 }}> </div>
            <div style={{ color: 'var(--light-fg)', alignSelf: 'center', fontSize: 'var(--small-font-size)', marginTop: '8px' }}>
              {mimeReqCount===1? Object.keys(shortMimeTypes)[0]
              : Object.keys(shortMimeTypes).map(k => (<div>
                  {shortMimeTypes[k]==='json'? (
                    <input type='radio' name='request_body_type' value={shortMimeTypes[k]} onChange={this.onMimeTypeChange} checked={true} style={{ margin: '0 0 0 8px'}} />
                  )
                  :(
                    <input type='radio' name='request_body_type' value={shortMimeTypes[k]} onChange={this.onMimeTypeChange} style={{margin: '0 0 0 8px'}} />
                  )}
                </div>))}
            </div>
          </div>
          <div id="tab_example" className={cx(stylett["tab-content"], stylef["col"])} style={styleExample} >
            {textareaExampleHtml[this.state.mimeTypeSelected]}
          </div>
          <div id="tab_model" className={cx(stylett["tab-content"], stylef["col"])} style={styleModel}>
            <SchemaTree data={reqSchemaTree}></SchemaTree>
          </div>
        </div>)
      }</React.Fragment>)


  }

}
