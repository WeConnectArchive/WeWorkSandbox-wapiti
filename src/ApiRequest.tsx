import * as React from 'react';
import marked from 'marked';
import DangerousHTML from 'react-dangerous-html';
import { schemaToModel, getTypeInfo,  generateExample, removeCircularReferences} from './utils/common-utils.js';
import {TagInput} from './TagInput';
import {SchemaTree} from './SchemaTree';
import './CommonStyles.css'
import './ApiRequest.css'
  
interface ApiRequestProps {
    specType: string;
    selectedServer: string;
    method: string;
    path: string;
    parameters: any[];
    requestBody: any;
    accept: string;
};

interface ApiRequestState {
   exampleTabActive: boolean; 
   mimeTypeSelected: string;
};

export class ApiRequest extends React.Component<ApiRequestProps, ApiRequestState> {

  constructor(props: ApiRequestProps) {
    super(props);
    this.state = { exampleTabActive: true, mimeTypeSelected: 'none' };
  }

  render() {
    return (
    <div className="request col regular-font request-panel">
      <div className="title">REQUEST</div>
      {this.inputParametersTemplate('path')}
      {this.inputParametersTemplate('query')}
      {this.requestBodyTemplate()}
      {this.inputParametersTemplate('header')}
      {this.inputParametersTemplate('cookie')}
    </div>
   );
  }

  inputParametersTemplate(paramType){
    let title ="";
    let filteredParams= this.props.parameters? this.props.parameters.filter( param => param.in === paramType ):[];
    if (filteredParams.length === 0 ){
      return "";
    }
    if (paramType==='path'){ title = "PATH PARAMETERS"}
    else if (paramType==='query'){ title = "QUERY-STRING PARAMETERS"}
    else if (paramType==='header'){ title = "REQUEST HEADERS"}
    else if (paramType==='cookie'){ title = "COOKIES"}


    const tableRows : any[] = [];
    for (const param of filteredParams) {
      if (!param.schema){
        continue;
      }
      let paramSchema = getTypeInfo(param.schema);
      if (paramSchema === undefined) return;
      let inputVal='';
      if (param.example){
        if (param.example==='0' || param.example === 0){
          inputVal='0'
        }
        else{
          inputVal = param.example;
        }
      }
      else{
         inputVal = paramSchema.default;
      }

      tableRows.push((<>
      <tr>
        <td style={{ minWidth: '100px'}}>
          <div className="param-name">
            {param.required && <span style={{ color: 'orangered' }}>*</span>}{param.name}
          </div>
          <div className="param-type">
          {paramSchema.type==='array' ? paramSchema.array
          : paramSchema.type + (paramSchema.format ? '\u00a0' + paramSchema.format : '')}
          </div>
        </td>
        <td style={{ minWidth: '100px'}}>
          {paramSchema.type === 'array' ? (
          <TagInput
            theClassName="request-param"
            theStyle={{ width: '100%', fontSize: 'calc(var(--small-) + 1px)', background: 'var(--input-bg)', lineHeight: '13px' }}
            dataPtype={paramType}
            dataPname={param.name}
            dataArray="true"
            placeholder="add-multiple\u23ce"
          ></TagInput>)
          :(<input type="text" spellCheck={false} style={{ width: '100%' }} className="request-param"
            value={inputVal}/>)}
        </td>
        <td>
          <div className="param-constraint"><>
            {paramSchema.constrain && (<>{paramSchema.constrain}<br/></>)}
            {paramSchema.allowedValues && (<>{paramSchema.allowedValues}</>)}
          </></div>
        </td>
      </tr>
      {param.description && (
        <tr>
          <td style={{ border: 'none' }}>

          </td>
          <td colSpan={2} style={{ border: 'none', marginTop: '0', padding: '0 5px' }}>
            <span className="m-markdown-small"><DangerousHTML html={marked(param.description)} /></span>
          </td>
        </tr>
        )}
       </>
      ))
    }

    return (<>
    <div className="table-title top-gap">{title}</div>
    <div style={{ display: 'block', overflowX: 'auto',  maxWidth: '100%' }}>
      <table className="m-table" style={{ width: '100%', wordBreak: 'break-word' }}>
        {tableRows}
      </table>
    </div>
    </>
    )
  }

  activateTab = (e) => {
    this.setState({ exampleTabActive: e.target.id === 'tab_example', mimeTypeSelected: this.state.mimeTypeSelected });
  }

  onMimeTypeChange = (e) => {
    this.setState({ exampleTabActive: this.state.exampleTabActive, mimeTypeSelected: e.target.value });
  }

  requestBodyTemplate(){

    if(!this.props.requestBody){
      return '';
    }
    if (Object.keys(this.props.requestBody).length === 0){
      return '';
    }

    let mimeReqCount=0;
    let shortMimeTypes={};
    let bodyDescrHtml = this.props.requestBody.description? (<div className="m-markdown"> <DangerousHTML html={marked(this.props.requestBody.description)} /></div>):'';
    let textareaExampleHtml:any[] = [];
    var formDataHtml: any;
    const formDataTableRows: any[] = [];
    let isFormDataPresent   = false;
    let reqSchemaTree="";

    let content = this.props.requestBody.content;

    for(let mimeReq in content ) {
      if (this.state.mimeTypeSelected === 'none') {
        this.setState({mimeTypeSelected: mimeReq, exampleTabActive: this.state.exampleTabActive })
      }
      // do not change shortMimeTypes values, they are referenced in other places
      if (mimeReq.includes('json')){shortMimeTypes[mimeReq]='json';}
      else if (mimeReq.includes('xml')){shortMimeTypes[mimeReq]='xml';}
      else if (mimeReq.includes('text/plain')){shortMimeTypes[mimeReq]='text';}
      else if (mimeReq.includes('form-urlencoded')){shortMimeTypes[mimeReq]='form-urlencoded';}
      else if (mimeReq.includes('multipart/form-data')){shortMimeTypes[mimeReq]='multipart-form-data';}

      let mimeReqObj = content[mimeReq];

      if (mimeReq.includes('json') || mimeReq.includes('xml') || mimeReq.includes('text/plain')){
        try {
          //Remove Circular references from RequestBody json-schema
          mimeReqObj.schema = JSON.parse(JSON.stringify(mimeReqObj.schema, removeCircularReferences()));
        }
        catch{
          console.error("Unable to resolve circular refs in schema", mimeReqObj.schema);
          return;
        }
        reqSchemaTree = schemaToModel(mimeReqObj.schema,{});
        let reqExample    = generateExample(
          mimeReqObj.schema? mimeReqObj.schema.examples:'',
          mimeReqObj.schema? mimeReqObj.schema.example:'',
          mimeReqObj.schema,
          mimeReq, "text"
        );
        textareaExampleHtml[mimeReq] = (<textarea
            className={'textarea mono request-body-param '+ shortMimeTypes[mimeReq]}
            spellCheck={false}
            style={{ resize: 'vertical', display: shortMimeTypes[mimeReq]==='json'?'block':'none' }}
            value={reqExample[0].exampleValue} />)
      }
      else if (mimeReq.includes('form') || mimeReq.includes('multipart-form')){
        isFormDataPresent = true;
        for (const fieldName in mimeReqObj.schema.properties)  {
          const fieldSchema = mimeReqObj.schema.properties[fieldName];
          const fieldType = fieldSchema.type;
          const arrayType = fieldSchema.type==='array'?fieldSchema.items.type:'';
          formDataTableRows.push((<>
          <tr>
            <td style={{minWidth: '100px' }}>
              <div className="param-name">{fieldName}</div>
              <div className="param-type">
              {fieldType==='array' ? '{fieldType} of {arrayType}'
              : fieldType + (fieldSchema.format ? '\u00a0' + fieldSchema.format : '')}
              </div>
            </td>
            <td style={{ minWidth: '100px' }}>
              {fieldType === 'array'? (
              <TagInput 
                theClassName="request-form-param"
                theStyle={{ width: '100%', fontSize: 'calc(var(--title-font-size) + 1px)', background: 'var(--input-bg)', lineHeight: '13px' }}
                dataPtype={fieldType}
                dataPname={fieldName}
                dataArray="true"
                placeholder="add-multiple\u23ce"
              ></TagInput>)
              :(<input
                spellCheck={false}
                type="{fieldSchema.format==='binary'?'file':'text'}"
                style={{ width: '100%' }} className="request-form-param"
                />
              )}

            </td>
            <td>
              <div className="param-constraint"></div>
            </td>
          </tr>
          {fieldSchema.description?(
            <tr>
              <td style={{ border: 'none' }}></td>
              <td colSpan={2} style={{ border: 'none', marginTop:0, padding: '0 5px' }}>
                <span className="m-markdown-small"><DangerousHTML html={marked(fieldSchema.description)} /></span>
              </td>
            </tr>):''}
           </>
        ))
        }


        formDataHtml = (
        <form className="{shortMimeTypes[mimeReq]}"> {/* onSubmit={{ event.preventDefault() }}> */}
          <table style={{ width: '100%' }} className="m-table">
            {formDataTableRows}
          </table>
        </form>);
      }
      mimeReqCount++;
    }

    let bodyTitle = "";
    if (this.props.specType === 'grpc') {
      bodyTitle = this.props.requestBody.objType
      textareaExampleHtml['protobuf'] = (<textarea
            className="textarea mono request-body-param json"
            spellCheck={false}
            style={{ resize: 'vertical', display: 'block' }}
            value={JSON.stringify(this.props.requestBody.mock, undefined, 2)} />)
      reqSchemaTree = this.props.requestBody.schema;
      shortMimeTypes['protobuf']={};
      if (this.state.mimeTypeSelected === 'none') {
        this.setState({mimeTypeSelected: 'protobuf', exampleTabActive: this.state.exampleTabActive })
      }
      mimeReqCount = 1;
    } else {
      bodyTitle = (isFormDataPresent?'FORM':'BODY') + ' DATA '+(this.props.requestBody.required?'(required)':'')
    }

    let styleExample=this.state.exampleTabActive?{ flex: 1 }:{ flex: 1, display: 'none' }
    let styleModel=this.state.exampleTabActive?{ flex: 1, display: 'none' }:{ flex: 1 }

    return (<>
      <div className={'table-title top-gap ' + (isFormDataPresent?'form_data':'body_data')}> {bodyTitle} </div>
      {bodyDescrHtml}
      {isFormDataPresent?({formDataHtml})
        :(
        <div className="tab-panel col" style={{borderWidth: '0',  minHeight: '200px' }}>
          <div id="tab_buttons" className="tab-buttons row" onClick={this.activateTab}>
            <button className={'tab-btn'+(this.state.exampleTabActive?' active':'')} id="tab_example">EXAMPLE </button>
            <button className={'tab-btn'+(this.state.exampleTabActive?'':' active')} id="tab_model">MODEL</button>
            <div style={{ flex:1 }}> </div>
            <div style={{ color: 'var(--light-fg)', alignSelf: 'center', fontSize: 'var(--small-font-size)', marginTop: '8px' }}>
              {mimeReqCount===1? Object.keys(shortMimeTypes)[0]
              : Object.keys(shortMimeTypes).map(k => (<>
                  {shortMimeTypes[k]==='json'? (
                    <input type='radio' name='request_body_type' value={shortMimeTypes[k]} onChange={this.onMimeTypeChange} checked={true} style={{ margin: '0 0 0 8px'}} />
                  )
                  :(
                    <input type='radio' name='request_body_type' value={shortMimeTypes[k]} onChange={this.onMimeTypeChange} style={{margin: '0 0 0 8px'}} />
                  )}
                </>))}
            </div>
          </div>
          <div id="tab_example" className="tab-content col" style={styleExample} >
            {textareaExampleHtml[this.state.mimeTypeSelected]}
          </div>
          <div id="tab_model" className="tab-content col" style={styleModel}>
            <SchemaTree data={reqSchemaTree}></SchemaTree>
          </div>
        </div>)
      }</>)


  }

}
