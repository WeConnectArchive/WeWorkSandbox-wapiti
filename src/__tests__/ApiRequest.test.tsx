import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { ApiRequest } from '../ApiRequest';

const theParams = [{
  name: "limit",
  in: "query",
  description: "How many items to return at one time (max 100)",
  required: "false",
  schema: {
     type: "integer",
     format: "int32"
  }
}]

const theBody = {
      description: "description",
      content: {
        "application/json": {
           schema:{
             type: "object",
             required: ["id","name"],
             properties: {
               id: {
                 type: "integer",
                 format: "int64"
               },
               name: {
                 type: "string"
               },
               tag: {
                 type: "string"
               }
             }
           }
        }
      }
}

describe('ApiRequest', () => {
  it('renders with rest GET', () => {
    let wrapper = shallow(<ApiRequest
        accept=""
        method="GET"
        parameters={theParams}
        path="/test"
        selectedServer=""
        specType="rest"
      ></ApiRequest>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('renders with rest POST', () => {
    let wrapper = shallow(<ApiRequest
        accept=""
        method="POST"
        requestBody={theBody}
        path="/test"
        selectedServer=""
        specType="rest"
      >></ApiRequest>
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });


});
