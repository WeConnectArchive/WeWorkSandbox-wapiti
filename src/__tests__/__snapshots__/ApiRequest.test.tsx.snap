// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiRequest renders with rest GET 1`] = `
<div
  className="request col regular-font request-panel"
>
  <div
    className="title"
  >
    REQUEST
  </div>
  <div
    className="table-title top-gap"
  >
    QUERY-STRING PARAMETERS
  </div>
  <div
    style={
      Object {
        "display": "block",
        "maxWidth": "100%",
        "overflowX": "auto",
      }
    }
  >
    <table
      className="m-table"
      style={
        Object {
          "width": "100%",
          "wordBreak": "break-word",
        }
      }
    >
      <tr>
        <td
          style={
            Object {
              "minWidth": "100px",
            }
          }
        >
          <div
            className="param-name"
          >
            <span
              style={
                Object {
                  "color": "orangered",
                }
              }
            >
              *
            </span>
            limit
          </div>
          <div
            className="param-type"
          >
            integer int32
          </div>
        </td>
        <td
          style={
            Object {
              "minWidth": "100px",
            }
          }
        >
          <input
            className="request-param"
            spellCheck={false}
            style={
              Object {
                "width": "100%",
              }
            }
            type="text"
            value=""
          />
        </td>
        <td>
          <div
            className="param-constraint"
          />
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "border": "none",
            }
          }
        />
        <td
          colSpan={2}
          style={
            Object {
              "border": "none",
              "marginTop": "0",
              "padding": "0 5px",
            }
          }
        >
          <span
            className="m-markdown-small"
          >
            <DangerousHTML
              className=""
              html="<p>How many items to return at one time (max 100)</p>
"
              tagName="div"
            />
          </span>
        </td>
      </tr>
    </table>
  </div>
</div>
`;

exports[`ApiRequest renders with rest POST 1`] = `
<div
  className="request col regular-font request-panel"
>
  <div
    className="title"
  >
    REQUEST
  </div>
  <div
    className="table-title top-gap body_data"
  >
     
    BODY DATA 
     
  </div>
  <div
    className="m-markdown"
  >
     
    <DangerousHTML
      className=""
      html="<p>description</p>
"
      tagName="div"
    />
  </div>
  <div
    className="tab-panel col"
    style={
      Object {
        "borderWidth": "0",
        "minHeight": "200px",
      }
    }
  >
    <div
      className="tab-buttons row"
      id="tab_buttons"
      onClick={[Function]}
    >
      <button
        className="tab-btn active"
        id="tab_example"
      >
        EXAMPLE 
      </button>
      <button
        className="tab-btn"
        id="tab_model"
      >
        MODEL
      </button>
      <div
        style={
          Object {
            "flex": 1,
          }
        }
      >
         
      </div>
      <div
        style={
          Object {
            "alignSelf": "center",
            "color": "var(--light-fg)",
            "fontSize": "var(--small-font-size)",
            "marginTop": "8px",
          }
        }
      >
        application/json
      </div>
    </div>
    <div
      className="tab-content col"
      id="tab_example"
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <div
      className="tab-content col"
      id="tab_model"
      style={
        Object {
          "display": "none",
          "flex": 1,
        }
      }
    >
      <SchemaTree
        data={
          Object {
            "id*": "integer int64~|~",
            "name*": "string~|~",
            "tag": "string~|~",
          }
        }
      />
    </div>
  </div>
</div>
`;
